piper/buttondialog.py

```diff
@@ -102,13 +128,112 @@ def _init_mapping_page(self, buttons):
             if self._button_mapping > 0 and button == self._button:
                 self.combo_mapping.set_active_id(key)
 
+        self._keystroke.bind_property("accel", self.label_keystroke, "accelerator")
+        self._keystroke.bind_property("accel", self.label_preview, "accelerator")
+        if self._button_mapping == 0 and self._button.type == "key":
+            keys = self._button.key
+            self._keystroke.from_evdev(keys[0], keys[1:])
+
     def _get_button_key_and_name(self, button):
         if button.type != "unknown":
             name = self._BUTTON_TYPE_TO_DESC[button.type]
         else:
             name = _("Button {} click").format(button.index)
         return str(button.index + 1), name  # Logical buttons are 1-indexed.
 
+    def _grab_seat(self):
+        # Grabs the keyboard seat. Returns True on success, False on failure.
+        # Gratefully copied from GNOME Control Center's keyboard panel.
+        window = self.get_window()
+        if window is None:
+            return False
+        display = window.get_display()
+        seats = display.list_seats()
+        if len(seats) == 0:
+            return False
+        device = seats[0].get_keyboard()
+        if device is None:
+            return False
+        if device.get_source == Gdk.InputSource.KEYBOARD:
+            pointer = device.get_associated_device()
+            if pointer is None:
+                return False
+        else:
+            pointer = device
+        status = pointer.get_seat().grab(window, Gdk.SeatCapabilities.KEYBOARD,
+                                         False, None, None, None, None)
+        if status != Gdk.GrabStatus.SUCCESS:
+            return False
+        self._grab_pointer = pointer
+        self.grab_add()
+        return True
+
+    def _release_grab(self):
+        # Releases a previously grabbed keyboard seat, if any.
+        if self._grab_pointer is None:
+            return
+        self._grab_pointer.get_seat().ungrab()
+        self._grab_pointer = None
+        self.grab_remove()
+
+    def _set_keystroke(self):
+        # Sets the current action type to "key" and caches the keystroke in a
+        # ratbagd compatible form, then restores the overview page and releases
+        # the grab.
+        self._action_type = "key"
+        self._key_mapping = self._keystroke.to_evdev()
+        self.stack_mapping.set_visible_child_name("overview")
+        self._release_grab()
+
+    def do_key_press_event(self, event):
+        # Overrides Gtk.Widget's standard key press event callback, so we can
+        # capture the pressed buttons in capture mode. Gratefully copied from
+        # GNOME Control Center's keyboard panel.
+        # Don't process key events when we're not in capture mode.
+        if self.stack_mapping.get_visible_child_name() == "overview":
+            return Gtk.Widget.do_key_press_event(self, event)
+
+        # TODO: remove this workaround when libratbag removes its keycode
```

Comment:
sorry, last 30s comment before I run off: you keep your own state of down/up and replay that in the same sequence. if you never saw a key up, that key is still down.